### PR TITLE
Refactor for easier deploys of new versions. Adds jemalloc 5.0.1

### LINKED ISSUE #2: Upgrade to 4.0.4

### LINKED ISSUE #7: New Heroku-16 stack available
Not sure if we actually need to change anything for this or not

### LINKED ISSUE #13: No heroku-16 build
I have:

```
$  heroku buildpacks
1. https://github.com/heroku/heroku-buildpack-apt
2. https://github.com/mojodna/heroku-buildpack-jemalloc.git#v3.6.0
3. heroku/ruby
```

When I deploy, the build phase crashes:

```
[...]
remote: -----> jemalloc app detected
remote: -----> Vendoring binaries
remote:        Fetching https://s3.amazonaws.com/mojodna-heroku/heroku-16/jemalloc-3.6.0-1.tar.gz
remote: 
remote: gzip: stdin: not in gzip format
remote: tar: Child returned status 1
remote: tar: Error is not recoverable: exiting now
remote:  !     Push rejected, failed to compile jemalloc app.
remote: 
remote:  !     Push failed
[...]
```

This file doesn't exist: https://s3.amazonaws.com/mojodna-heroku/heroku-16/jemalloc-3.6.0-1.tar.gz

Everything is fine if I don't pin to `v3.6.0`.

Is there a reason 3.6.0 is still shown as a fallback option? Is it still supported?

### PR DESCRIPTION
**Pull request by gaffneyc**
This PR rewrites a large chunk of the build system and how the buildpack is installed and configured. The goals with this PR are to make using jemalloc easier for folks and to reduce the time and work required to release new versions.

The build system has been changed to build specific versions for multiple Heroku stacks against the official Heroku docker images. This means building a new version is a single command `make VERSION=5.0.1` followed by uploading the build artifacts to a release on GitHub.

There has been some confusion lately around setting LD_PRELOAD especially since the revision changed with 5.0.0. This introduces JEMALLOC_ENABLED which can configure LD_PRELOAD automatically. This has the added benefit of being opt-in rather than opt-out. People have the option to either set this or use `jemalloc.sh` on a per-dyno basis.

Versions can now be switched by setting JEMALLOC_VERSION and pushing a new code change / recompiling the slug. The previous incarnation of versions required there to be a specific tag on GitHub with the version string. Now, anyone using the buildpack will always have the most up to date installer script and can switch between versions as long as there is a build available. This also means no longer needing to make code changes to release a new version of jemalloc.

There is some configuration that needs to happen if or when this PR is merged. It uses a different organization for releases where there is a tag based on name of the Heroku stack (heroku-16 and cedar-14). I'm not aware of a way to create a release on GitHub that only contains uploaded binaries (i.e. doesn't include any source) so I created a `release-master` that is an orphaned commit with no code on it. See [the releases on my fork](https://github.com/gaffneyc/heroku-buildpack-jemalloc/releases) for a reference. I'm happy to configure the releases here to match it or be responsible for updating new builds on my fork going forward.

Finally, this work was sponsored by [Dead Man's Snitch](https://deadmanssnitch.com) (a cron monitoring service), where we deployed 5.0.1 and have seen some really great results on both Puma and Sidekiq. 

Fixes #2
Fixes #7
Fixes #13
Closes #11 
Closes #14

### CODE REVIEW

**Comment by jonekdahl** [comment]
Having support for heroku-16 would be great as we had to remove the buildpack from our app when upgrading to heroku-16 recently.

**Comment by gaffneyc** [comment]
@jonekdahl take a look at [my fork](https://github.com/gaffneyc/heroku-buildpack-jemalloc) which is the basis for this PR. It's usable today and support heroku-16

**Comment by jonekdahl** [comment]
Any chance of this getting merged? I am still using @gaffneyc's fork and it works well for us (on heroku-16 with JEMALLOC_VERSION=3.6.0) but I would prefer to use the upstream repo.

**Comment by gaffneyc** [comment]
@jonekdahl I plan to keep my fork up to date and maintained since we've deployed it to a number of apps in the last month and switching back would be a lot of work.

**Comment by nateberkopec** [comment]
OK, I'm back in the open-source world.

@gaffneyc Anything that needs to be changed/updated here before I sit down and take a look?

**Comment by gaffneyc** [comment]
@nateberkopec No, this is good to go. It depends on a specific setup for the releases tab before being merged (which I mention in the description).

I'm planning to keep my fork maintained, if you or mojodna don't have capacity then I'm happy to take on ownership of this.

**Comment by sethfitz** [comment]
@gaffneyc I don't have capacity, so thanks for volunteering! @nateberkopec has admin rights on this repo (pretty sure), so I'll defer to him on how to continue supporting this.

**Comment by jayelkaake** [comment]
I think this should be merged - worked great for us! Thank you sir!

**Comment by gaffneyc** [comment]
For anyone following this, [jemalloc 5.1.0](https://github.com/jemalloc/jemalloc/releases/tag/5.1.0) was released yesterday. I've added it as well as builds for the upcoming Heroku-18 stack to my fork. It looks like 5.1.0 is a recommended upgrade by the jemalloc developers and we've just deployed it to our staging environment. Hopefully if all goes well I'll make 5.1.0 the default in a month or so.

**Comment by nateberkopec** [comment]
LGTM, suggest adding @gaffneyc as maintainer and let him merge/release this.

**Comment by sethfitz** [comment]
@gaffneyc I've just invited you as a collaborator. Thanks!!!

**Comment by gaffneyc** [comment]
Awesome, thanks. I've done a bit of work on my fork since this PR. I'll see about getting this merged, releases put together, and other changes added in here.

**Comment by gaffneyc** [comment]
Sorry, I haven't had time to come back and get this merged. The problem is that it's a rebased and edited version of an old version of what's on master on [my branch](https://github.com/gaffneyc/heroku-buildpack-jemalloc) right now. I'll try to schedule some time this week to make it happen.

**Comment by czj** [comment]
Thanks a lot for your work @gaffneyc il would be awesome :)

**Comment by bf4** [comment]
@gaffneyc @nateberkopec I can't tell what's blocking this PR from being merged. Does it need to be broken up into smaller PRs? I'm happy to help, just don't want to get started without checking in.

**Comment by gaffneyc** [comment]
@bf4 This is blocked by me having time to do it. It's based off of [my fork](https://github.com/gaffneyc/heroku-buildpack-jemalloc) but it's both lagged behind master on my fork and was a modified rebase to keep the downloads coming from this repo. To get the PR up to date I would need to go through that process again from my master branch and add the releases. It's been hard to prioritize the time since my fork is working for what we need.

If you need what's in this PR today then I would say just use my fork. If you have time to help get it merged, going through the rebase process and opening a new PR would help a bunch.

**Comment by bf4** [comment]
If I understand what you mean, is that this PR is based on a different branch than what your'e using now, and that since the pr source branch, you've made changes to your master which should be part of any PR ? https://github.com/gaffneyc/heroku-buildpack-jemalloc/compare/gaffneyc:upgrade-jemalloc...master Asking because the PR we're commenting on looked fine to me, though I admittedly haven't run it. I'll try cherry-picking some stuff.

**Comment by bf4** [comment]
@gaffneyc I took your master minus the sponsorship commit, since I figure that's a separate discussion, and merged this repos master into it. Was a clean merge. https://github.com/mojodna/heroku-buildpack-jemalloc/pull/19

**Comment by wongy91** [comment]
how do i verify if jemalloc is running on heroku/dokku (im running RoR)

**Comment by bf4** [comment]
@wongy91 
> how do i verify if jemalloc is running on heroku/dokku (im running RoR)

That's *really* off topic.  I'm pretty sure that the right way to get help for this buildpack, assuming you're using it, is to 1) look for an existing issue 2) create a new one showing what you've tried, what you see, and what you expected

**Comment by dbackeus** [comment]
Currently it's about a 50/50 split in popularity between this original repo and @gaffneyc's fork looking at heroku buildpacks list:
https://elements.heroku.com/search/buildpacks?q=jemalloc

Would be nice to reduce confusion and fragmentation of usage by deprecating one of the repos. Most straight-forward way would seem to be to get things merged into original repo rather than switching to the fork?

**Comment by gaffneyc** [comment]
**tldr;** Migrate to [gaffneyc/heroku-buildpack-jemalloc](http://github.com/gaffneyc/heroku-buildpack-jemalloc) if you want newer versions of jemalloc. Or stick with this buildpack if things are stable for you.

Since opening this PR almost 11 months ago I haven't had much time to revisit it. Even with #19 there is still work to be done to finish the migration and an inherent risk to anyone depending on this buildpack that a major change like this could break their applications on the next deploy.

I'm committed to maintaining my fork since we depend on it for [Dead Man's Snitch](https://deadmanssnitch.com) and several client projects at [Collective Idea](https://collectiveidea.com). With limited time the idea of maintaining two branches of the same code isn't very enticing and it's shown over the last year as I've been more responsive on my fork to issues.

Given the risk to people using this buildpack and the reality of the time I have available, I'm closing this PR (sorry :/) and recommending that people migrate to my fork as they're able to. This means stability for anyone who's happy with the current buildpack but it also means the migration will be opt-in, giving teams the ability to choose when and how they handle it.
